connect/src/com/telenor/connect/headerenrichment/MobileDataFetcher.java

```diff
@@ -0,0 +1,140 @@
+        }
+
+        int statusCode = webResourceResponse.getStatusCode();
+        if (statusCode < 200 || statusCode >= 300) {
+            return null;
+        }
+
+        try {
+            InputStream inputStream = webResourceResponse.getData();
+            String response = IOUtils.toString(inputStream, "UTF-8");
+            inputStream.close();
+            return response;
+        } catch (IOException e) {
+            return null;
+        }
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+    public static WebResourceResponse fetchWebResourceResponse(String originalUrl, boolean allowedToToggleNetworkToUse) {
+        String newUrl = originalUrl;
+        int attempts = 0;
+        Network interfaceToUse = HeLogic.getCellularNetwork();
+        if (interfaceToUse == null) {
+            return null;
+        }
+        do {
+            try {
+                HttpURLConnection connection
+                        = (HttpURLConnection)interfaceToUse.openConnection(new URL(newUrl));
+                connection.setInstanceFollowRedirects(false);
+                connection.connect();
+                int responseCode = connection.getResponseCode();
+                attempts += 1;
+                if (responseCode != HTTP_SEE_OTHER
+                        && responseCode != HTTP_MOVED_TEMP
+                        && responseCode != HTTP_MOVED_PERM) {
+                    // Rely on the WebView to close the input stream when finished fetching data
+                    return new WebResourceResponse(
+                            connection.getContentType(),
+                            connection.getContentEncoding(),
+                            connection.getResponseCode(),
+                            connection.getResponseMessage(),
+                            getHeadersAsCommaJoinedList(connection.getHeaderFields()),
+                            connection.getInputStream());
+                }
+                newUrl = connection.getHeaderField("Location");
+                // Close the input stream, but do not disconnect the connection as its socket might
+                // be reused during the next request.
+                connection.getInputStream().close();
+            } catch (final IOException e) {
+                return null;
+            }
+            if (allowedToToggleNetworkToUse) {
+                interfaceToUse = shouldFetchThroughCellular(newUrl)
+                        ? HeLogic.getCellularNetwork()
+                        : HeLogic.getDefaultNetwork();
+            }
+        } while (attempts <= MAX_REDIRECTS_TO_FOLLOW_FOR_HE);
+        return null;
+    }
+
+    private static Map<String, String> getHeadersAsCommaJoinedList(Map<String, List<String>> headerFields) {
+        Map<String, String> result = new HashMap<>();
+        for (String header : headerFields.keySet()) {
+            List<String> strings = headerFields.get(header);
+            if (strings == null || strings.isEmpty()) {
```

Comment:
Not sure how much string processing is done and if it makes sense, but there is a method com.google.common.base.Strings.isNullOrEmpty(). You've most likely seen it in the connect repository.

Reply:
Could use `TextUtils.isEmpty(string)`, Android's built-in version. I think my argument is that I find it faster and safer to write out the null check, since it's short, and since there are hundred different string null check util methods one can use. If I did use that, someone not familiar with that method would have to look it up to verify it does a null check, although I guess you could reasonably assume it does. What do you think?

Reply:
To me, Strings.isNullOrEmpty() is very descriptive. That night be because I'm used to it. However, considering it checks for two things in one method (null or empty), it's more readable than two conditionals. 
I have no strong opinions, two conditionals work just fine :)

Reply:
Or maybe I'm mixing some things up here... This appears to be checking that a list is not null and not empty, not a string, as I thought, when I wrote the first comment.

So neither Strings.isNullOrEmpty(String) nor TextUtils.isEmpty(String) would work in this case.

Reply:
I see that now, good point :O Perhaps the parameter should be renamed to reflect that, e.g. `stringList` or `headerList`.

Reply:
Changed: 1e09acf